invoices/interface.go

```diff
@@ -163,3 +163,46 @@ type InvoiceSlice struct {
 // CircuitKey is a tuple of channel ID and HTLC ID, used to uniquely identify
 // HTLCs in a circuit.
 type CircuitKey = models.CircuitKey
+
+// InvoiceUpdaterContext is the context that's passed to the methods of the
+// InvoiceUpdater interface. It contains all the information needed to update
+// the invoice in the database.
+type InvoiceUpdaterContext struct {
+	// invoice holds the updated invoice. It is always set regardless
+	// of the type of update.
+	Invoice *Invoice
+
+	// updateTime is the timestamp used for the update.
+	UpdateTime time.Time
+
+	// settledSetIDs holds the set of set IDs that were settled as part of
+	// this update. This is only set for StoreAddHtlcsUpdate.
+	SettledSetIDs map[SetID]struct{}
+
+	// ampHtlcsUpdate holds the set of AMP HTLCs that were added or
+	// cancelled as part of this update. This is only set for
+	// StoreAddHtlcsUpdate and StoreCancelHtlcsUpdate.
+	AMPHTLCsUpdate map[SetID]map[models.CircuitKey]*InvoiceHTLC
+}
+
+// InvoiceUpdater is an interface to abstract away the details of updating an
+// invoice in the database. The methods of this interface are called when the
+// in-memory update of an invoice is complete, and the database needs to be
+// updated.
+type InvoiceUpdater interface {
```

Comment:
+1 for having most of the logic here in the `invoices` package

Then we have another layer between this and the DB, that's responsible for mapping the invoice updates into concrete DB operations.